zerver/migrations/0699_fix_user_role_system_groups.py

```diff
@@ -0,0 +1,191 @@
+ROLE_REALM_ADMINISTRATOR = 200
+ROLE_MODERATOR = 300
+ROLE_MEMBER = 400
+ROLE_GUEST = 600
+
+
+class SystemGroups:
+    FULL_MEMBERS = "role:fullmembers"
+    EVERYONE_ON_INTERNET = "role:internet"
+    OWNERS = "role:owners"
+    ADMINISTRATORS = "role:administrators"
+    MODERATORS = "role:moderators"
+    MEMBERS = "role:members"
+    EVERYONE = "role:everyone"
+    NOBODY = "role:nobody"
+
+
+def fix_system_group_memberships_based_on_role(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
+    """
+    Our ldap integration supports syncing the user's .role based on their
+    LDAP groups. It had a bug, updating only the .role value, without updating
+    the system group memberships for the user accordingly. Additionally,
+    the RealmAuditLog USER_ROLE_CHANGED entry was not created either.
+
+    This migration fixes the group memberships for users whose .role
+    doesn't match them, additionally also creating the missing RealmAuditLog
+    object.
+    """
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    Realm = apps.get_model("zerver", "Realm")
+    RealmAuditLog = apps.get_model("zerver", "RealmAuditLog")
+    NamedUserGroup = apps.get_model("zerver", "NamedUserGroup")
+    UserGroupMembership = apps.get_model("zerver", "UserGroupMembership")
+
+    def get_system_user_group_by_name(group_name: str, realm_id: int) -> Any:
+        return NamedUserGroup.objects.get(name=group_name, realm_id=realm_id, is_system_group=True)
+
+    def realm_user_count_by_role(realm: Any) -> dict[str, Any]:
+        """
+        Code copied from the implementation in zerver.lib.user_counts
+        """
+        ROLE_COUNT_HUMANS = "11"
+        ROLE_COUNT_BOTS = "12"
+
+        human_counts = {
+            str(ROLE_REALM_ADMINISTRATOR): 0,
+            str(ROLE_REALM_OWNER): 0,
+            str(ROLE_MODERATOR): 0,
+            str(ROLE_MEMBER): 0,
+            str(ROLE_GUEST): 0,
+        }
+        for value_dict in (
+            UserProfile.objects.filter(realm=realm, is_bot=False, is_active=True)
+            .values("role")
+            .annotate(Count("role"))
+        ):
+            human_counts[str(value_dict["role"])] = value_dict["role__count"]
+        bot_count = UserProfile.objects.filter(realm=realm, is_bot=True, is_active=True).count()
+        return {
+            ROLE_COUNT_HUMANS: human_counts,
+            ROLE_COUNT_BOTS: bot_count,
+        }
+
+    def fix_user_memberships_if_needed(user_profile: Any) -> None:
+        realm = user_profile.realm
+
+        guest_group = get_system_user_group_by_name(SystemGroups.EVERYONE, realm.id)
+        member_group = get_system_user_group_by_name(SystemGroups.MEMBERS, realm.id)
+        moderator_group = get_system_user_group_by_name(SystemGroups.MODERATORS, realm.id)
+        admin_group = get_system_user_group_by_name(SystemGroups.ADMINISTRATORS, realm.id)
+        owner_group = get_system_user_group_by_name(SystemGroups.OWNERS, realm.id)
```

Comment:
We should avoid doing these 5 database queries for every user; these should get passed as an object from the caller after being fetched for the realm once, or something.